### PR TITLE
core: add missing interrupt masking in get/set user map for LPAE

### PR DESCRIPTION
**Pull request by etienne-lms**
strange that we did saw issues before...
maybe i missed something.

The proposed fixes were found from a review of the use of `get_core_pos()` in optee_os.

### CODE REVIEW

**Comment by vchong** [comment]
Do we need to add the same to get user map function in `core_mmu_v7.c` as well?

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 1]
```diff
@@ -861,13 +867,17 @@ void core_mmu_get_user_map(struct core_mmu_user_map *map)
...
 }
 
 void core_mmu_set_user_map(struct core_mmu_user_map *map)
 {
 	uint64_t ttbr;
-	uint32_t daif = read_daif();
+	uint32_t daif;
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
 
+	daif = read_daif()
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu_lpae.c]: The daif stuff is interrupt masking so you can remove that now that we can use `thread_mask_exceptions()` instead.

Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu_lpae.c]: ok, thanks.

**Comment by etienne-lms** [comment]
Thanks for feedback. So you're right, these changes do not fix anything:
- core_mmu_get_user_map() is always called with interrupts masked.
- core_mmu_set_user_map() already masks interrupts (`read/write_daif()` in lpae, `thread_mask_exceptions` in 32bit mmu).

So i will push back this p-r with cleaning changes only:
- core_mmu_lpae.c: use `thread_mask_exceptions` instead of `read/write_daif`.
- assert in `core_mmu_get_user_map()` that foreign interrupts are masked.

**Comment by etienne-lms** [comment]
I tried to assert foreign interrupts are masked while in `get_core_pos()` but I ended in code i find not very nice. Here it is. If you think it's fine and `get_core_pos()` deserves such, i will add it to this P-R:

```patch
--- a/core/arch/arm/kernel/misc_a32.S
+++ b/core/arch/arm/kernel/misc_a32.S
@@ -31,13 +31,35 @@
 #include <arm32_macros.S>
 #include <kernel/unwind.h>
 #include <platform_config.h>
+#include <kernel/thread.h>
 
 /* Let platforms override this if needed */
 .weak get_core_pos_mpidr
 
+#ifdef CFG_TEE_CORE_DEBUG
+.section .rodata
+panic_get_core_pos_func:
+	.asciz "get_core_pos"
+#endif
+
 /* size_t get_core_pos(void); */
 FUNC get_core_pos , :
 UNWIND(	.fnstart)
+
+#ifdef CFG_TEE_CORE_DEBUG
+	mrs	r0, cpsr
+	tst	r0, #(THREAD_EXCP_FOREIGN_INTR << CPSR_F_SHIFT)
+	/*
+	 * this must be compliant with the panic generic routine:
+	 * __do_panic(__FILE__, __LINE__, __func__, str)
+	 */
+	moveq	r0, #0
+	moveq	r1, #0
+	ldreq	r2, =panic_get_core_pos_func
+	moveq	r3, #0
+	beq	__do_panic
+#endif
+
 	read_mpidr r0
 	b get_core_pos_mpidr
 UNWIND(	.fnend)
```

**Comment by jenswi-linaro** [comment]
You could use a C wrapper function for the assert part instead.

**Comment by jenswi-linaro** [comment]
I think adding an assert for masked interrupts in `get_core_pos()` would be an efficient way of catching this kind of errors.

**Comment by jforissier** [comment]
Eh, yes, all the more since Aarch64 has to be handled, too ;-)

**Comment by etienne-lms** [comment]
- `NDEBUG` or `CFG_TEE_CORE_DEBUG` ?
Both are linked. I find `CFG_TEE_CORE_DEBUG` more standard to optee.

> why don't you like it?
+ 
> You could use a C wrapper function for the assert part instead.

`get_core_pos()` may be called without a valid stack, hence I would rather not unconditionally push data to the stack from this function, especially at core boot (cold and warm boots). In such case, an assert (which should not occur since at boot interrupts are expected masked) will corrupt the core without printing the expected nice trace.

Maybe assert only when `get_core_pos()` is called from C (assert from a C definition of `get_core_pos`) ?

**Comment by jenswi-linaro** [comment]
Yes, only asserting when from C is fine.

**Comment by jforissier** [comment]
> Yes, only asserting when from C is fine.

Then the discussion about `NDEBUG` vs. `CFG_TEE_CORE_DEBUG` will be moot, I suppose ;-)

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 2]
```diff
@@ -770,6 +770,12 @@ bool core_mmu_user_mapping_is_active(void)
 #ifdef ARM32
 void core_mmu_get_user_map(struct core_mmu_user_map *map)
 {
+	/*
+	 * Foreign interrupts must be disabled before playing with current
+	 * core since we otherwise may be rescheduled to a different core.
+	 */
+	assert(thread_get_exceptions() & THREAD_EXCP_FOREIGN_INTR);
```
Comment (jforissier) [inline core/arch/arm/mm/core_mmu_lpae.c]: Not needed now that `get_core_pos()` has the assert

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 3]
```diff
@@ -854,6 +860,12 @@ void core_mmu_get_user_map(struct core_mmu_user_map *map)
 {
 	assert(user_va_idx != -1);
 
+	/*
+	 * Foreign interrupts must be disabled before playing with current
+	 * core since we otherwise may be rescheduled to a different core.
+	 */
+	assert(thread_get_exceptions() & THREAD_EXCP_FOREIGN_INTR);
```
Comment (jforissier) [inline core/arch/arm/mm/core_mmu_lpae.c]: Ditto.

**Comment by jenswi-linaro** [comment]
With Jeromes comments addressed:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/mm/core_mmu_lpae.c (hunk 2 above)
Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu_lpae.c]: right :)

**Comment by jforissier** [comment]
Crashes on HiKey (see HiKey auto builder):
```
D/TC:0 open_session:400 open entry point for pseudo ta "invoke_tests.pta"
E/TC:0 assertion 'thread_get_exceptions() & THREAD_EXCP_FOREIGN_INTR' failed at core/arch/arm/include/kernel/misc.h:45 <get_core_pos>
E/TC:0 Panic at core/kernel/assert.c:51 <_assert_break>
E/TC:0 Call stack:
E/TC:0  0x000000003f00a2b8 read_pc at optee_os/core/arch/arm/include/arm64.h:226
E/TC:0  0x000000003f014f48 __do_panic at optee_os/core/kernel/panic.c:53
E/TC:0  0x000000003f013acc _assert_break at optee_os/core/kernel/assert.c:51
E/TC:0  0x000000003f00dbb8 get_core_pos at optee_os/core/arch/arm/include/kernel/misc.h:45
E/TC:0  0x000000003f00dec8 core_mmu_find_table at optee_os/core/arch/arm/mm/core_mmu_lpae.c:645
E/TC:0  0x000000003f00bdfc core_mmu_unmap_pages at optee_os/core/arch/arm/mm/core_mmu.c:1390
E/TC:0  0x000000003f00f96c mobj_reg_shm_unmap at optee_os/core/arch/arm/mm/mobj.c:519
E/TC:0  0x000000003f00f9c0 mobj_reg_shm_free at optee_os/core/arch/arm/mm/mobj.c:375
E/TC:0  0x000000003f010d58 tee_entry_std at optee_os/core/arch/arm/tee/entry_std.c:580
E/TC:0  0x000000003f007c4c __thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread.c:601
E/TC:0  0x000000003f003380 thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread_a64.S:184
```

**Comment by jenswi-linaro** [comment]
Aha, good assert it seems. :-)

**Comment by jforissier** [comment]
https://github.com/OP-TEE/optee_os/pull/2064

**Pull request merged**